File: src/main/java/org/academy/api/client/gui/layout/Gravity.java
package org.academy.api.client.gui.framework.layout;

public final class Gravity {
    public static final int NO_GRAVITY = 0;

    public static final int AXIS_SPECIFIED = 0x0001;
    public static final int AXIS_PULL_BEFORE = 0x0002;
    public static final int AXIS_PULL_AFTER = 0x0004;
    public static final int AXIS_X_SHIFT = 0;
    public static final int AXIS_Y_SHIFT = 4;

    public static final int TOP = (AXIS_PULL_BEFORE | AXIS_SPECIFIED) << AXIS_Y_SHIFT;
    public static final int BOTTOM = (AXIS_PULL_AFTER | AXIS_SPECIFIED) << AXIS_Y_SHIFT;
    public static final int LEFT = (AXIS_PULL_BEFORE | AXIS_SPECIFIED) << AXIS_X_SHIFT;
    public static final int RIGHT = (AXIS_PULL_AFTER | AXIS_SPECIFIED) << AXIS_X_SHIFT;

    public static final int CENTER_VERTICAL = AXIS_SPECIFIED << AXIS_Y_SHIFT;
    public static final int CENTER_HORIZONTAL = AXIS_SPECIFIED << AXIS_X_SHIFT;
    public static final int CENTER = CENTER_VERTICAL | CENTER_HORIZONTAL;

    public static final int TOP_LEFT = TOP | LEFT;

    public static final int HORIZONTAL_GRAVITY_MASK = (AXIS_SPECIFIED | AXIS_PULL_BEFORE | AXIS_PULL_AFTER) << AXIS_X_SHIFT;
    public static final int VERTICAL_GRAVITY_MASK = (AXIS_SPECIFIED | AXIS_PULL_BEFORE | AXIS_PULL_AFTER) << AXIS_Y_SHIFT;

    public static final int FILL_HORIZONTAL = HORIZONTAL_GRAVITY_MASK;
    public static final int FILL_VERTICAL = VERTICAL_GRAVITY_MASK;
    public static final int FILL = FILL_HORIZONTAL | FILL_VERTICAL;

    public static final int RELATIVE_LAYOUT_DIRECTION = 0x00800000;

    public static final int START = RELATIVE_LAYOUT_DIRECTION | LEFT;
    public static final int END = RELATIVE_LAYOUT_DIRECTION | RIGHT;

    public static final int RELATIVE_HORIZONTAL_GRAVITY_MASK = START | END;

    public static final int TOP_RIGHT = TOP | RIGHT;
    public static final int BOTTOM_LEFT = BOTTOM | LEFT;
    public static final int BOTTOM_RIGHT = BOTTOM | RIGHT;

    public static final int CENTER_TOP = CENTER_HORIZONTAL | TOP;
    public static final int CENTER_BOTTOM = CENTER_HORIZONTAL | BOTTOM;
    public static final int CENTER_LEFT = CENTER_VERTICAL | LEFT;
    public static final int CENTER_RIGHT = CENTER_VERTICAL | RIGHT;

    private Gravity() {
    }
}

File: src/main/kotlin/org/academy/api/client/gui/layout/Gravity.kt
package org.academy.api.client.gui.layout

object Gravity {
    const val NO_GRAVITY: Int = 0

    const val AXIS_SPECIFIED: Int = 0x0001
    const val AXIS_PULL_BEFORE: Int = 0x0002
    const val AXIS_PULL_AFTER: Int = 0x0004
    const val AXIS_X_SHIFT: Int = 0
    const val AXIS_Y_SHIFT: Int = 4

    const val TOP: Int = (AXIS_PULL_BEFORE or AXIS_SPECIFIED) shl AXIS_Y_SHIFT
    const val BOTTOM: Int = (AXIS_PULL_AFTER or AXIS_SPECIFIED) shl AXIS_Y_SHIFT
    const val LEFT: Int = (AXIS_PULL_BEFORE or AXIS_SPECIFIED) shl AXIS_X_SHIFT
    const val RIGHT: Int = (AXIS_PULL_AFTER or AXIS_SPECIFIED) shl AXIS_X_SHIFT

    const val CENTER_VERTICAL: Int = AXIS_SPECIFIED shl AXIS_Y_SHIFT
    const val CENTER_HORIZONTAL: Int = AXIS_SPECIFIED shl AXIS_X_SHIFT
    const val CENTER: Int = CENTER_VERTICAL or CENTER_HORIZONTAL

    const val TOP_LEFT: Int = TOP or LEFT

    const val HORIZONTAL_GRAVITY_MASK: Int = (AXIS_SPECIFIED or AXIS_PULL_BEFORE or AXIS_PULL_AFTER) shl AXIS_X_SHIFT
    const val VERTICAL_GRAVITY_MASK: Int = (AXIS_SPECIFIED or AXIS_PULL_BEFORE or AXIS_PULL_AFTER) shl AXIS_Y_SHIFT

    const val FILL_HORIZONTAL: Int = HORIZONTAL_GRAVITY_MASK
    const val FILL_VERTICAL: Int = VERTICAL_GRAVITY_MASK
    const val FILL: Int = FILL_HORIZONTAL or FILL_VERTICAL

    const val RELATIVE_LAYOUT_DIRECTION: Int = 0x00800000

    const val START: Int = RELATIVE_LAYOUT_DIRECTION or LEFT
    const val END: Int = RELATIVE_LAYOUT_DIRECTION or RIGHT

    const val RELATIVE_HORIZONTAL_GRAVITY_MASK: Int = START or END

    const val TOP_RIGHT: Int = TOP or RIGHT
    const val BOTTOM_LEFT: Int = BOTTOM or LEFT
    const val BOTTOM_RIGHT: Int = BOTTOM or RIGHT

    const val CENTER_TOP: Int = CENTER_HORIZONTAL or TOP
    const val CENTER_BOTTOM: Int = CENTER_HORIZONTAL or BOTTOM
    const val CENTER_LEFT: Int = CENTER_VERTICAL or LEFT
    const val CENTER_RIGHT: Int = CENTER_VERTICAL or RIGHT

    fun  apply(){}
}